### PR TITLE
Fix redirects when updating personal details

### PR DESCRIPTION
**Pull request by TobyRet**
## Context

_The first in possibly, a large-ish number of PRs_

If you visit the review unsubmitted application form page and click change on any of the links, then either:

1. update the field and click save and continue 
or
2. click the backlink

You are redirected the review page for the section.

In both of these cases you should be redirected to the review submitted page 

## Changes proposed in this pull request

Use params (`&return-to=application-review`) to redirect the candidate back to the application review path when required.

## Guidance to review

- This PR deals with the personal details section only. Other PRs to follow for remaining sections.

## Link to Trello card

https://trello.com/c/0MxsfCi5/3720-bug-change-links-and-backlinks-are-not-working-correctly-from-the-review-unsubmitted-page

## Things to check

- [x] This code does not rely on migrations in the same Pull Request
- [ ] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [ ] API release notes have been updated if necessary
- [ ] Required environment variables have been updated [added to the Azure KeyVault](/docs/environment-variables.md#deploy-pipeline)

### CODE REVIEW

File: app/views/candidate_interface/application_form/_review.html.erb [hunk 1]
```diff
@@ -31,7 +31,7 @@
 <section class="govuk-!-margin-bottom-8">
   <h2 class="govuk-heading-m govuk-!-font-size-27"><%= t('section_groups.personal_details') %></h2>
   <h3 class="govuk-heading-m"><%= t('page_titles.personal_information') %></h3>
-  <%= render(CandidateInterface::PersonalDetailsReviewComponent.new(application_form: application_form, editable: editable, missing_error: missing_error)) %>
+  <%= render(CandidateInterface::PersonalDetailsReviewComponent.new(application_form: application_form, editable: editable, missing_error: missing_error, return_to_application_review: true)) %>
```
Comment (TobyRet) [inline app/views/candidate_interface/application_form/_review.html.erb:34]: This feels a bit gross - we have to pass a flag down to the `PersonalDetailsReviewPresenter` which is responsible for populating the `return_to` params in the change link as the PersonalDetailsReview component is used both on the application page and the application review page - but not sure if there is an easier way

File: app/controllers/candidate_interface/personal_details/languages_controller.rb [hunk 2]
```diff
@@ -21,13 +21,21 @@ def create
 
       def edit
         @languages_form = LanguagesForm.build_from_application(current_application)
+
+        @return_to = if redirect_back_to_application_review_page?
+                       application_review_path_and_params
+                     else
+                       { path: candidate_interface_personal_details_complete_path, params: {} }
+                     end
```
Comment (malcolmbaig) [inline app/controllers/candidate_interface/personal_details/languages_controller.rb:29]: Maybe we can put this `if` block behind a method in CandidateInterfaceController. Something like:

```
@return_to = return_to_after_edit(default: candidate_interface_personal_details_complete_path)
...

def return_to_after_edit(default:)
  if redirect_back_to_application_review_page?
    { path: candidate_interface_application_review_path, params: { 'return-to' => 'application-review' } }
  else
    { path: default, params: {} }
  end
end
```

(and then we can get rid of `application_review_path_and_params, which doesn't appear to be needed anywhere else)

File: app/controllers/candidate_interface/personal_details/languages_controller.rb [hunk 3]
```diff
@@ -21,13 +21,21 @@ def create
...
       end
 
       def update
         @application_form = current_application
         @languages_form = LanguagesForm.new(languages_params)
 
         if @languages_form.save(current_application)
+          return redirect_to candidate_interface_application_review_path if redirect_back_to_application_review_page?
```
Comment (malcolmbaig) [inline app/controllers/candidate_interface/personal_details/languages_controller.rb:37]: For each of these cases, could the return_to path that we resolve in the `edit` action be hidden fields in the form? Could potentially do just the one call here then:

`redirect_to params.fetch(return_to_path)`

I might be missing something, but I think that means you then don't need to handle the `return-to` param in the edit template.

File: app/controllers/candidate_interface/personal_details/languages_controller.rb (hunk 2 above)
Comment (TobyRet) [reply app/controllers/candidate_interface/personal_details/languages_controller.rb:29]: Nice - like that idea 👍

File: app/views/candidate_interface/application_form/_review.html.erb (hunk 1 above)
Comment (malcolmbaig) [reply app/views/candidate_interface/application_form/_review.html.erb:34]: The additional arg seems fine to me, and could always be switched from a boolean to a string/symbol later on if there are ever more than two possible return paths.

I wonder, though, if this is a good opportunity to collapse the presenter into PersonalDetailsReviewComponent? Doesn't seem to be much point in keeping them separate and might be easier to follow.

**Review by malcolmbaig — COMMENTED** [review]
This is looking good, a few suggestions left inline.

File: app/controllers/candidate_interface/personal_details/languages_controller.rb (hunk 3 above)
Comment (TobyRet) [reply app/controllers/candidate_interface/personal_details/languages_controller.rb:32]: That's an interesting one - I had a play around with this idea but I think it added more complexity. When a candidate visits the application review page and decides to update a field (i.e. by clicking 'Change') we need to past the 'return_to' as query params. By sticking with query params the whole way through it may make the code a little cleaner rather than having a mix of query params and hidden fields (?).

File: app/views/candidate_interface/application_form/_review.html.erb (hunk 1 above)
Comment (TobyRet) [reply app/views/candidate_interface/application_form/_review.html.erb:34]: Had a look into this also, and from what I can see, the reason for the separation may have been to make testing easier (the presenter makes good use of dependency injection allowing for multiple edge cases to be tested). Methinks collapsing will result in result in quite a cumbersome / complex test suite.

**Comment by TobyRet** [comment]
@malcolmbaig - realised I forgot to fix the contact details backlinks/redirects - have added as a separate commit - https://github.com/DFE-Digital/apply-for-teacher-training/pull/5273/commits/5004eb37aa736ee005eceab9cdbec55eac653417
